### PR TITLE
Transfer learning

### PR DESCRIPTION
**Pull request by AlexDBlack**
***DO NOT MERGE***

### CODE REVIEW

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/FrozenLayerTest.java [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+//        assertEquals(modelNow.getLayer(2).conf(), notFrozen.getLayer(0).conf());  //Equal, other than names
+//        assertEquals(modelNow.getLayer(3).conf(), notFrozen.getLayer(1).conf());  //Equal, other than names
+
+        //Check: forward pass
+        INDArray outNow = modelNow.output(randomData.getFeatures());
+        INDArray outNotFrozen = notFrozen.output(asFrozenFeatures);
+        assertEquals(outNow, outNotFrozen);
+
+        for( int i=0; i<5; i++ ){
+            notFrozen.fit(new DataSet(asFrozenFeatures,randomData.getLabels()));
```
Comment (agibsonccc) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/FrozenLayerTest.java]: spaciing. (i < 5)

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/FrozenLayerTest.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+
+    @Test
+    public void cloneMLNFrozen() {
+
+        DataSet randomData = new DataSet(Nd4j.rand(10,4),Nd4j.rand(10,3));
+
+        NeuralNetConfiguration.Builder overallConf = new NeuralNetConfiguration.Builder().learningRate(0.1).optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT).updater(Updater.SGD).activation(Activation.IDENTITY);
+        MultiLayerNetwork modelToFineTune = new MultiLayerNetwork(overallConf.list()
```
Comment (agibsonccc) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/FrozenLayerTest.java]: there's this neat thing called an "enter key" :D

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/FrozenLayerTest.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+                        .nIn(2).nOut(3)
+                        .build())
+                .layer(1, new org.deeplearning4j.nn.conf.layers.OutputLayer.Builder(LossFunctions.LossFunction.MCXENT)
+                        .activation(Activation.SOFTMAX)
+                        .nIn(3).nOut(3)
+                        .build()).build(),Nd4j.hstack(modelToFineTune.getLayer(2).params(),modelToFineTune.getLayer(3).params()));
+
+        int i = 0;
+        while (i<5) {
+            notFrozen.fit(new DataSet(asFrozenFeatures,randomData.getLabels()));
```
Comment (agibsonccc) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/FrozenLayerTest.java]: spacing.

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+        NeuralNetConfiguration.Builder overallConf = new NeuralNetConfiguration.Builder().learningRate(0.001).optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT).updater(Updater.SGD);
+        /*
+                        inCentre        inRight
+                            |               |
+                        denseCentre0        |
+                            |               |
+             ---------  denseCentre1     denseRight0
+            /               |               |
+        subsetLeft(0-3) denseCentre2 ----mergeRight
```
Comment (agibsonccc) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java]: needs more pacman ascii art.

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        ComputationGraph modelToTune = new ComputationGraph(conf);
+        modelToTune.init();
+
+        INDArray [] features = new INDArray[] {Nd4j.rand(10,10),Nd4j.rand(10,2)};
+        INDArray [] labels = new INDArray[] {Nd4j.rand(10,6),Nd4j.rand(10,4),Nd4j.rand(10,5)};
+
+        //Instead of manually setting these three vertices - the comp graph .setFeatureExtractor should have a way of saying freeze from inCentre to denseCentre2
+        //and should trace the path from inCentre to denseCenter2 and freeze only those (not freeze the mergeRight etc etc)
+        modelToTune.getVertex("denseCentre0").setLayerAsFrozen();
+        modelToTune.getVertex("denseCentre1").setLayerAsFrozen();
```
Comment (agibsonccc) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java]: this is an interesting idea. Why not define this in the builder?

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+        ComputationGraph graph = new ComputationGraph(conf);
+        graph.init();
+
+
+        int[] topologicalOrder = graph.topologicalSortOrder();
+        org.deeplearning4j.nn.graph.vertex.GraphVertex[] vertices = graph.getVertices();
+
+        for( int i=0; i<topologicalOrder.length; i++ ){
+            org.deeplearning4j.nn.graph.vertex.GraphVertex v = vertices[topologicalOrder[i]];
```
Comment (agibsonccc) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java]: spacing.

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        ComputationGraph graph2 = new TransferLearning.GraphBuilder(graph)
+                .fineTuneConfiguration(new FineTuneConfiguration.Builder()
+                        .learningRate(2e-2).build())
+                .setFeatureExtractor("C")
+                .build();
+
+        boolean cFound = false;
+        Layer[] layers = graph2.getLayers();
+
+        for(Layer l : layers){
```
Comment (agibsonccc) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java]: spacing.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/FineTuneConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    protected Boolean minimize;
+    protected GradientNormalization gradientNormalization;
+    protected Double gradientNormalizationThreshold;
+    protected LearningRatePolicy learningRatePolicy;
+    protected Double lrPolicyDecayRate;
+    protected Double lrPolicySteps;
+    protected Double lrPolicyPower;
+    protected ConvolutionMode convolutionMode;
+
+    protected Boolean pretrain;
```
Comment (agibsonccc) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/FineTuneConfiguration.java]: Note that if you want default values then specify those int his builder here.

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/FrozenLayerTest.java (hunk 2 above)
Comment (eraly) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/layers/FrozenLayerTest.java]: O Really :D ?

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java (hunk 5 above)
Comment (eraly) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java]: A user should not have to do this manually. The user sets this in the builder with .setFeatureExtractor...

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/FineTuneConfiguration.java (hunk 8 above)
Comment (eraly) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/FineTuneConfiguration.java]: We want to check for nulls/nans as unset which is why it's not initialized here. Part of the logic of how to be able to override the configurations

**Pull request closed without merging**